src/init/ResourceStoreInitializer.ts

```diff
@@ -0,0 +1,71 @@
+import { DataFactory } from 'n3';
+import { RepresentationMetadata } from '../ldp/representation/RepresentationMetadata';
+import type { ResourceIdentifier } from '../ldp/representation/ResourceIdentifier';
+import { getLoggerFor } from '../logging/LogUtil';
+import type { ResourceStore } from '../storage/ResourceStore';
+import { TEXT_TURTLE } from '../util/ContentTypes';
+import { NotFoundHttpError } from '../util/errors/NotFoundHttpError';
+import { ensureTrailingSlash } from '../util/PathUtil';
+import { generateResourceQuads } from '../util/ResourceUtil';
+import { guardedStreamFrom } from '../util/StreamUtil';
+import { PIM, RDF } from '../util/UriConstants';
+import { toNamedNode } from '../util/UriUtil';
+import { Initializer } from './Initializer';
+import namedNode = DataFactory.namedNode;
+
+/**
+ * Initializes ResourceStores by creating a root container if it didn't exist yet.
+ */
+export class ResourceStoreInitializer extends Initializer {
+  protected readonly logger = getLoggerFor(this);
+  private readonly baseId: ResourceIdentifier;
+  private readonly store: ResourceStore;
+
+  public constructor(baseUrl: string, store: ResourceStore) {
+    super();
+    this.baseId = { path: ensureTrailingSlash(baseUrl) };
+    this.store = store;
+  }
+
+  public async handle(): Promise<void> {
+    if (!await this.hasRootContainer()) {
+      await this.createRootContainer();
+    }
+  }
+
+  /**
+   * Verify if a root container already exists in a ResourceStore.
+   */
+  private async hasRootContainer(): Promise<boolean> {
+    try {
+      const result = await this.store.getRepresentation(this.baseId, {});
+      this.logger.debug(`Existing root container found at ${this.baseId.path}`);
+      result.data.destroy();
+      return true;
+    } catch (error: unknown) {
+      if (!(error instanceof NotFoundHttpError)) {
+        throw error;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Create a root container in a ResourceStore.
+   */
+  private async createRootContainer(): Promise<void> {
+    const metadata = new RepresentationMetadata(this.baseId);
+    metadata.addQuads(generateResourceQuads(namedNode(this.baseId.path), true));
+
+    // Make sure the root container is a pim:Storage
```

Comment:
Short summary: Actually not sure if it is needed 😄 

The spec says this:
> When a DELETE request targets storage’s root container or its associated ACL resource, the server MUST respond with the 405 status code.

This was originally (wrongly) interpreted as meaning that the root container can not be deleted, which is why we had the root check. Adding a `pim:Storage` type to the root container is to still have the same behaviour as we had before (not allowing the root container to just be deleted), but in a way that matches the spec.

But actually there is nothing saying that the root container should be a storage container, so this could be dropped actually. Although I don't think it's bad to have this, since you probably don't want to delete your root.

Reply:
Why is that _wrongly_?

I think we need to drop `pim:Storage` (not up to us to put it there), but preserve the 405 on deletion attempts.

Reply:
> Why is that _wrongly_?

Because it says _storage_ root container, not root container. This was mentioned initially in https://github.com/solid/community-server/issues/381#issuecomment-737277297 .

A storage is something specifically defined in the spec: https://solid.github.io/specification/#storage

Unless I misinterpreted that.

Reply:
Right; let’s keep it as in this PR for now. It’s a valid interpretation.

Reply:
…but do add to the comment 🙂